Get low scoped access token

kgangineni opened the pull request:
### Summary of changes

 - Get access token using client id
 - Use it in `PaypalWebCheckoutClient`
 - Token is not used right now, added todo to be used in getting app switch eligibility
 ### Checklist

 - [ ] Added a changelog entry

### Authors
> List GitHub usernames for everyone who contributed to this pull request.

-

kgangineni committed:
⏺ Add OAuth2 client token fetching with suspend PayPalWebCheckoutClient

  - Implement FetchClientToken class for OAuth2 client credentials flow
  - Add headers parameter to APIRequest for Basic auth support
  - Convert PayPalWebCheckoutClient.start() to suspend function
  - Integrate token fetching before launching web checkout
  - Add comprehensive error handling for token fetch failures

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/FetchClientToken.kt:
```diff
@@ -0,0 +1,44 @@
+        val httpResponse = restClient.send(apiRequest)
+        if (httpResponse.isSuccessful) {
+            val jsonObject = JSONObject(httpResponse.body ?: "")
+            return jsonObject.getString("access_token")
```
Could we use `jsonObject.optString()` here instead? I believe the regular `getString()` method will throw when `access_token` doesn't exist.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/FetchClientToken.kt:
```diff
@@ -0,0 +1,44 @@
+            val jsonObject = JSONObject(httpResponse.body ?: "")
+            return jsonObject.getString("access_token")
+        } else {
+            throw httpResponse.run { buildError(status, body) }
```
Could we return a `FetchClientResult.Failure()` type here instead of throwing?

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/FetchClientToken.kt:
```diff
@@ -0,0 +1,44 @@
+import org.json.JSONObject
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+class FetchClientToken(
```
```suggestion
class AuthenticationSecureTokenServiceAPI(
```
This name would fit our naming convention if we want to be 100% consistent with the other `API` suffixed classes.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/FetchClientToken.kt:
```diff
@@ -0,0 +1,44 @@
+    private val coreConfig: CoreConfig,
+    private val restClient: RestClient = RestClient(configuration = coreConfig),
+) {
+    suspend operator fun invoke(): String {
```
```suggestion
    suspend fun createToken(): String {
```
We mostly use the functor pattern for UseCase type classes. I am interested in this pattern. If we decide to go this route, I'd like for all of the other `API` suffixed classes to migrate to this pattern so we can keep it consistent.

kgangineni replied:
renaming to `getToken()` because same token is returned after it's created

sshropshire replied:
It isn't a new token every time? Also, I went with "create" in the naming prefix because the underlying HTTP method is a `POST`.

sshropshire commented on CorePayments/src/test/java/com/paypal/android/corepayments/api/FetchClientTokenUnitTest.kt:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    @Test
+    fun `invoke() returns access token from successful response`() = runTest {
```
While it's unlikely, I think we should add a test for when `access_token` is missing. I think right now it'll throw a `JSONException`.

kgangineni committed:
⏺ Add callback-based start function with PayPalWebCheckoutCallback interface

  - Create PayPalWebCheckoutCallback fun interface following CardApproveOrderCallback pattern
  - Add overloaded start() function accepting PayPalWebCheckoutCallback

KunJeongPark commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/AuthenticationSecureTokenServiceAPI.kt:
```diff
@@ -0,0 +1,66 @@
+    private val coreConfig: CoreConfig,
+    private val restClient: RestClient = RestClient(configuration = coreConfig),
+) {
+    suspend fun getClientToken(): GetClientTokenResult {
```
Are these called client tokens?

KunJeongPark replied:
@sshropshire Unrelated, but wee put the UpdateClientConfigAPI` class in `PayPalWebCheckout `module for this draft PR:`
https://github.com/paypal/paypal-android/pull/314 and iOS version also, but I think we should move it to `CorePayments` as well, don't you think? I think originally we thought that this would be only for PayPalWebCheckout module, but since we have make that call everywhere, should live in CorePayments, right?

KunJeongPark replied:
just one field I think would be dynamic.

sshropshire replied:
Yeah `UpdateClientConfigAPI` sounds like it should live in the core module, that makes sense to me.

kgangineni replied:
Are these called client tokens?
> @KunJeongPark technically it's access token, may be go full descriptive? `getLowScopedAccessToken`

KunJeongPark replied:
that sounds good to me. What do you think @sshropshire ?

sshropshire replied:
I do like `getLowScopedAccessToken()` it feels like the most expressive name.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/AuthenticationSecureTokenServiceAPI.kt:
```diff
@@ -0,0 +1,65 @@
+        val requestBody = "grant_type=client_credentials&response_type=token"
+
+        val headers = mutableMapOf(
+            "Authorization" to "Basic ${coreConfig.clientId.base64encoded()}",
```
```suggestion
            "Authorization" to "Basic ${coreConfig.clientId.base64encoded()}:",
```
nit: does this work without the colon at the end? I remember we used to do that when we had LSATs in the past.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/APIClientError.kt:
```diff
@@ -91,4 +92,16 @@ object APIClientError {
+        errorDescription = errorDescription.orEmpty(),
+        correlationId = correlationId,
+        reason = reason
+    )
```
🧶 : do we need this? Why don't we call the `PayPalSDKError` constructor directly in this scenario.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/AuthenticationSecureTokenServiceAPI.kt:
```diff
@@ -0,0 +1,65 @@
+    private val coreConfig: CoreConfig,
+    private val restClient: RestClient = RestClient(configuration = coreConfig),
+) {
+    suspend fun getLowScopedAccessToken(): GetClientTokenResult {
```
```suggestion
    suspend fun createLowScopedAccessToken(): GetClientTokenResult {
```
nit: Since we're technically creating a token rather than retrieving one that already exists.

sshropshire commented on CorePayments/src/main/java/com/paypal/android/corepayments/api/GetClientTokenResult.kt:
```diff
@@ -0,0 +1,21 @@
+import com.paypal.android.corepayments.PayPalSDKError
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+sealed class GetClientTokenResult {
```
```suggestion
sealed class CreateLowScopedAccessTokenResult {
```

🧶 : to align result name with function method name.

sshropshire reviewed (approved):
Couple take or leave 🧶 's but overall LGTM
